src/addrman.h

```diff
@@ -417,20 +440,48 @@ class CAddrMan
         }
         nTried -= nLost;
 
-        // Deserialize positions in the new table (if possible).
+        // Store positions in the new table buckets to apply later (if possible).
+        std::map<int, int> entryToBucket; // Represents which entry belonged to which bucket when serializing
+
         for (int bucket = 0; bucket < nUBuckets; bucket++) {
             int nSize = 0;
             s >> nSize;
             for (int n = 0; n < nSize; n++) {
                 int nIndex = 0;
                 s >> nIndex;
                 if (nIndex >= 0 && nIndex < nNew) {
-                    CAddrInfo &info = mapInfo[nIndex];
-                    int nUBucketPos = info.GetBucketPosition(nKey, true, bucket);
-                    if (nVersion == 1 && nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && vvNew[bucket][nUBucketPos] == -1 && info.nRefCount < ADDRMAN_NEW_BUCKETS_PER_ADDRESS) {
-                        info.nRefCount++;
-                        vvNew[bucket][nUBucketPos] = nIndex;
-                    }
+                    entryToBucket[nIndex] = bucket;
+                }
+            }
+        }
+
+        uint256 supplied_asmap_version;
+        if (m_asmap.size() != 0) {
+            supplied_asmap_version = SerializeHash(m_asmap);
+        }
+        uint256 serialized_asmap_version;
+        if (nVersion > 1) {
+            s >> serialized_asmap_version;
+        }
+
+        for (int n = 0; n < nNew; n++) {
+            CAddrInfo &info = mapInfo[n];
+            int bucket = entryToBucket[n];
+            int nUBucketPos = info.GetBucketPosition(nKey, true, bucket);
+            if (nVersion == 2 && nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && vvNew[bucket][nUBucketPos] == -1 &&
```

Comment:
Why is this `nVersion == 2` check required? This means that anyone upgrading to v0.20 will have a lot of their new table data lost, even if they're not using an asmap. Without this check, we'd still make sure not to place entries in the wrong new table because of the asmap version check.

On code readability/style, it'd make a lot more sense to extract this check to above the loop and comment it:

```c++
// New table positions can only be retained if the file is the right format and
// was constructed with the same asmap.
const bool compatible_file = nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && serialized_asmap_version == supplied_asmap_version
```

Reply:
More generally, what's the plan for updating the asmap? Presumably it's something that could change fairly frequently (i.e. at least as often as we release major versions of Bitcoin Core). If that's the case and the asmap is changed every major version, are we happy with the idea that people's new tables will lose data at every major upgrade?

Reply:
>This means that anyone upgrading to v0.20 will have a lot of their new table data lost, even if they're not using an asmap. 

You are right, I think this was a mistake. I was probably thinking about downgrading `nVersion`, so the check should be `nVersion <= 2`.
I agree with your comment suggestion, but why you separate those two other checks? It still makes sure that "file is in the right format".

Reply:
>More generally, what's the plan for updating the asmap?
TBD

>If that's the case and the asmap is changed every major version, are we happy with the idea that people's new tables will lose data at every major upgrade?

Depends on the answer to the previous question at least :)
Note that what's lost is just the position of records in the table, and presumably a few of records from buckets with high collisions? So I wouldn't be *very* worried (in the context of my previous message as well).
It's not like we're loosing many ADDR records here.

Reply:
> what's lost is just the position of records in the table, and presumably a few of records from buckets with high collisions

A new address can appear in up to 8 different buckets if it's been gossipped to us from different peers. If it gets rebucketed during file load, it'll only be placed in 1 bucket (or 0, if there's a collision). That seems undesirable.

Reply:
You are right.